model/pipeline/src/main/resources/org/apache/beam/model/pipeline/pipeline.json

```diff
@@ -0,0 +1,320 @@
+        "windowMappingFn": { "type": "string" }
+      }
+    },
+
+     // The parameters to a ParDo transform and further metatdata
+    "ParDoPayload": {
+      "type": "object",
+      "additionalProperties": false,
+
+      "properties": {
+        // The DoFn for this ParDo
+        "doFn": { "type": "string" },
+
+        // Whether the ParDo requires access to the current window (this may force runners to perform
+        // additional work prior to executing this ParDo)
+        "requiresWindowAccess": { "type": "boolean" },
+
+        // A map from PCollection id to the specification for how to materialize and read it
+        // as a side input
+        "sideInputs": {
+          "type": "object",
+          "additionalProperties": { "$ref" : "#/definitions/SideInput" }
+        }
+      }
+    },
+
+    // The parameters to a Read transform and further metadata.
+    //
+    // TODO: decide what metadata should be here in runner-readable form vs what is part of the
+    // UDF.
+    "ReadPayload": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+
+        // The ID of the user-definable source fn
+        "source": { "type": "string" },
+
+        "isBounded": { "type": "boolean" },
+        "doesNotRequireSplitting": { "type": "boolean" },
+        "producesSortedKeys": { "type": "boolean" }
+      }
+    },
+
+    // The parameters to a Combine transform and further metadata.
+    "CombinePayload": {
+      "type": "object",
+      "additionalProperties": false,
+
+      "properties": {
+        // The ID of the user-defined CombineFn
+        "combineFn": { "type": "string" }
+      }
+    },
+
+    "WindowIntoPayload": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "windowingStrategy": { "type": "string" }
+      }
+    },
+
+    // A transform node contains pointers to its inputs and outputs, whether
+    // or not it is composite. If it is part of a composite, it contains a pointer
+    // to its parent.
+    "TransformNode": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+
+        // The stable name of the transform application node.
+        "label": { "type": "string" },
+
+        // Whether or not it is composite, it contains a "kind" string that uniquely
+        // identifies what it computes. This is a stable, cross-language string, that
+        // can be used for replacement-based implementation and also to parse the node
+        // to a concrete language-specific class.
+        //
+        // A stable, cross-language identifier identifying what the transform computes.
+        // This is to be used for replacement-based implementation, whether it is a primitive
+        // or composite. Examples include "GroupByKey", "Sum", "ParDo". We should likely
+        // adopt a standard namespacing convention, such as URIs.
+        "urn":  { "type": [ "null", "string" ] },
```

Comment:
Our options are to have randomly generated junk, null, or force every PTransform to have a URN. I'm not so sure we should force user transforms to have a URN. No runner will know what to do with them anyhow.